support `for [await]...of` statements

alexlamsl opened the pull request.

alexlamsl commented:
@kzc unrelated, but TIL this `vm` bug on Node.js v4~6:
```js
var a = 0;
try {
    (function async() {});
} finally {
    console.log(delete a);
}
(function() {
    () => {};
});
```
```sh
$ cat test.js | node
false

$ echo 'require("vm").runInNewContext(require("fs").readFileSync("test.js","utf8"),{console:console})' | node
true
```

kzc commented:
ES6 support is coming along nicely.  Just classes and import/export remaining to be implemented?

As for the node 4/6 vm bug - the global `var` in `test.js` would be wrapped in an IIFE in the sandbox, would it not? As such one would expect the `delete` of the local `var` to return false in that case. Pretty obscure even by our standards!

alexlamsl commented:
> ES6 support is coming along nicely. Just classes and import/export remaining to be implemented?

If you say so 👻 

> As for the node 4/6 vm bug - the global `var` in test.js would be wrapped in an IIFE in the sandbox, would it not?

Only when `toplevel` is specified, which shows up in `Infinity`, `NaN` & `undefined`.

kzc commented:
On the topic of obscure specification details... when `AST_Super` is implemented in conjunction with classes, be aware it can also be used in object literals:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/super#using_super.prop_in_object_literals

```js
var obj1 = {
  method1() {
    console.log('method 1');
  }
}

var obj2 = {
  method2() {
    super.method1();
  }
}

Object.setPrototypeOf(obj2, obj1);
obj2.method2(); // logs "method 1"
```

You may recall we derived `AST_Super` from `AST_This` in uglify-es due to their similar behaviors.

alexlamsl commented:
Truth be told, it's more difficult to implement `ufuzz` for this feature than to support it in the rest of `uglify-js` 😅 &minus; the usual case of these new features come with a whole new set of run-time exceptions.

kzc commented:
No doubt. It is very difficult to generate plausible scenarios in order to fuzz these features that are rife with usage constraints.

alexlamsl commented:
... and that trend isn't slowing down any time soon:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator#no_chaining_with_and_or_or_operators

kzc commented:
I didn't know about that one. It should have been a warning at best.

In the event that syntax is still parseable without the parentheses and non-ambiguous, uglify could allow it and introduce the parens in the output.  Then at least fuzzing won't be impeded.

alexlamsl commented:
Speaking of which, we do accept `for (foo of bar, baz)` just like the `for ... in` case (even though the spec thinks otherwise for the former), and always prints `for (foo of (bar, baz))`

kzc commented:
Another example... did you already mention this one?

This for-in is weird, but fine:
```
$ echo 'x=9;var y=7,z=[4,2];for(var x=y in z) console.log(x,z[x]);' | node
0 4
1 2
```
```
$ echo 'x=9;var y=7,z=[4,2];for(var x=y in z) console.log(x,z[x]);' | uglify-js -c toplevel | node
0 4
1 2
```
This for-of is not:
```
$ echo 'x=9;var y=7,z=[4,2];for(var x=y of z) console.log(x,z[x]);' | uglify-js -c toplevel
x=9;var x,z=[4,2];for(x of z)console.log(x,z[x]);
```
```
$ echo 'x=9;var y=7,z=[4,2];for(var x=y of z) console.log(x,z[x]);' | uglify-js -c toplevel | node
4 undefined
2 undefined
```
```
$ echo 'x=9;var y=7,z=[4,2];for(var x=y of z) console.log(x,z[x]);' | node
[stdin]:1
x=9;var y=7,z=[4,2];for(var x=y of z) console.log(x,z[x]);
                            ^^^
SyntaxError: for-of loop variable declaration may not have an initializer.
```

alexlamsl commented:
Oops indeed I forgot about this other weird syntax with `for ... in`

(Your example also highlights a potential weakness in `join_vars`...)

alexlamsl commented:
@kzc TIL:
```js
$ cat test.js
function to_hex(t) {
    return [].map.call(t, function(c) {
        return "_" + c.charCodeAt(0).toString(16);
    }).join("");
}

var s = "\ud87e\udc00";
console.log("length", s.length);
console.log("split", s.split("").map(to_hex));
console.log("for... in:");
for (var i in s)
    console.log(i, to_hex(s[i]));
console.log("for... of:");
for (var c of s)
    console.log(to_hex(c));
```
```sh
$ cat test.js | node
length 2
split [ '_d87e', '_dc00' ]
for... in:
0 _d87e
1 _dc00
for... of:
_d87e_dc00
```

kzc commented:
I see what they did there - they tried to right a historical wrong. But now it's inconsistent with the classic pattern which, like for/in, yields raw surrogates:
```
for (var i = 0; i < s.length; i++) {
    var value = s[i];
    console.log(to_hex(value));
}
```
Thanks. These are always amusing.

alexlamsl commented:
Stumbled upon this while looking for how column position is defined in source maps. Between code points of Unicode and code units of USCS/UTF-16/UTF-8 (Java/ECMAScript/sourcemap) I think I'll just leave the status quo be for now...

On another note, I don't seem to be able to find an _O(1)_ equivalent of `.length` for code points, which makes me think `.codePointAt()` could be more expensive than `.charCodeAt()` 🤔

kzc commented:
I don't think it's possible to calculate a utf-8 code point length in less than O(n) time. Hmm... perhaps with SSE instructions it could be slightly faster... sure enough there's [this](https://lemire.me/blog/2020/10/20/ridiculously-fast-unicode-utf-8-validation/). I don't know whether the JS engines store the code point length in the internal string object. It might be rarely used and not worth the overhead for typical use.

alexlamsl commented:
Indeed &minus; and if you want to do code-point `line` & `column` positioning, you are pretty much forced to `.split(/\n/)` the input source string first, otherwise the conversion would get hairy real quick.

Not to mention sourcemap stated succinctly in [its specification](https://sourcemaps.info/spec.html):
> #### Encoding
> 
> For simplicity, the character set encoding is always UTF-8.

If by that they meant UTF-8 code units, then it's even more expensive than `.codePointAt()` AFAICS 🤨

Of course this is only used for debugging, so you are supposed to enlist TOP500 to alleviate any performance concerns...

kzc commented:
>> For simplicity, the character set encoding is always UTF-8.
> If by that they meant UTF-8 code units

The column offsets in the source map spec are ambiguous. Although there's an emerging consensus to use code points...

https://github.com/evanw/esbuild/blob/master/CHANGELOG.md#0613

alexlamsl commented:
> Although there's an emerging consensus to use code points...

Erm... except in your link it says:

> Change source map column offsets from bytes to UTF-16 code units

"UTF-16 code units" refers to `.charCodeAt()` (which incidentally is what `uglify-js` is using since day#1 👻)

kzc commented:
Ah, okay. I missed that UTF-16 qualifier for code units.
